src/common/src/array/map_array.rs

```diff
@@ -0,0 +1,294 @@
+        }
+    }
+
+    impl PartialEq for MapValue {
+        fn eq(&self, _other: &Self) -> bool {
+            unreachable!("map is not comparable. Such usage should be banned in frontend.")
+        }
+    }
+
+    impl PartialOrd for MapValue {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    impl Ord for MapValue {
+        fn cmp(&self, _other: &Self) -> Ordering {
+            unreachable!("map is not comparable. Such usage should be banned in frontend.")
+        }
+    }
+
+    impl PartialEq for MapRef<'_> {
+        fn eq(&self, _other: &Self) -> bool {
+            unreachable!("map is not comparable. Such usage should be banned in frontend.")
+        }
+    }
+
+    impl Ord for MapRef<'_> {
+        fn cmp(&self, _other: &Self) -> Ordering {
+            unreachable!("map is not comparable. Such usage should be banned in frontend.")
+        }
+    }
+
+    impl PartialOrd for MapRef<'_> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+}
+
+impl Debug for MapValue {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.as_scalar_ref().fmt(f)
+    }
+}
+
+impl Display for MapValue {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.as_scalar_ref().write(f)
+    }
+}
+
+impl MapValue {
+    pub fn from_list_entries(list: ListValue) -> Self {
+        if cfg!(debug_assertions) {
+            // validates list type is valid
+            _ = MapType::from_list_entries(list.data_type());
+        }
+        // TODO: validate the values is valid
+        MapValue(list)
+    }
+
+    pub fn from_kv(keys: ListValue, values: ListValue) -> Self {
+        if cfg!(debug_assertions) {
+            assert_eq!(
+                keys.len(),
+                values.len(),
+                "keys: {keys:?}, values: {values:?}"
+            );
+            let unique_keys = keys.iter().unique().collect_vec();
+            assert!(
+                unique_keys.len() == keys.len(),
+                "non unique keys in map: {keys:?}"
+            );
+            assert!(!unique_keys.contains(&None), "null key in map: {keys:?}");
+        }
+
+        let len = keys.len();
+        let key_type = keys.data_type();
+        let value_type = values.data_type();
+        let struct_array = StructArray::new(
+            MapType::struct_type_for_map(key_type, value_type),
+            vec![keys.into_array().into_ref(), values.into_array().into_ref()],
+            Bitmap::ones(len),
+        );
+        MapValue(ListValue::new(struct_array.into()))
+    }
+}
+
+/// A map is just a slice of the underlying struct array.
+///
+/// XXX: perhaps we can make it `MapRef<'a, 'b>(ListRef<'a>, ListRef<'b>);`.
+/// Then we can build a map ref from 2 list refs without copying the data.
+/// Currently it's impossible.
```

Comment:
Good point. We can make it an `enum`.

Reply:
Why an `enum`? I think `(ListRef<'a>, ListRef<'b>)` can completely replace it. Note that `ListRef` is actually a generalized ref over an Array, not specific to list type. 😄

Btw, in `arrow-rs`, the `Array` types are actually ref types. And so `ListArray::iter`'s elements are just `ListArray`, and `MapArray` iterates over `StructArray`. 

But our arrays are owned types, which made it harder to do `MapRef` properly here.

Reply:
You're pretty right. 😄

Reply:
Hmmm, it's indeed not that simple, as we cannot get the underlying `(ListRef<'_>, ListRef<'_>)` from a `ListRef<'_>`.

It's `Arc<ArrayImpl>` vs `&'a ArrayImpl` thing 🤡

Reply:
Let's delay this due to the complexity https://github.com/risingwavelabs/risingwave/issues/17863

Reply:
Don't quite get that. This works for me:

```rust
fn make<'a>(a: ListRef<'a>) -> (ListRef<'a>, ListRef<'a>) {
    let c = &*a.array.as_struct().children;
    let array1 = &*c[0];
    let array2 = &*c[1];
    (
        ListRef {
            array: array1,
            start: a.start,
            end: a.end,
        },
        ListRef {
            array: array2,
            start: a.start,
            end: a.end,
        },
    )
}
```

Reply:
Oh, you are right. I misinterpreted the problem. 🤣

~~The real problem is that we cannot do sth like `ListValue -> ListRef (tmp variable) -> (ListRef, ListRef)`~~

Because I did
```rust
impl<'a> ListRef<'a> {
	fn as_map_kv(&self) -> (ListRef<'_>, ListRef<'_>) {}
}

# SHould be
impl<'a> ListRef<'a> {
	fn as_map_kv(self) -> (ListRef<'a>, ListRef<'a>) {}
}
```

Haven't played with lifetime for a long time!